### DISCUSSION TITLE
[java] AvoidAccessibilityAlteration: What is this rule about?

### DISCUSSION THREAD

**Discussion by linusjf** [pmd/pmd#3010] [Q&A]
**Description:**
[AvoidAccessibilityAlteration](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#avoidaccessibilityalteration)

Can you explain what's the purpose of the above rule?

Why is the rule restricted only to constructors and not methods and fields?

Why is the rule enforced despite the presence of a PrivilegedAction block? 

Surely, if reflection code is encapsulated within a PrivilegedAction then the accessibility violation is deliberate i.e., the coder intends to do exactly that (break encapsulation).

Spotbugs has a similar rule that enforces use of accessibility code within a doPrivilegedAction block. 

[DP_DO_INSIDE_DO_PRIVILEGED](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#dp-method-invoked-that-should-be-only-be-invoked-inside-a-doprivileged-block-dp-do-inside-do-privileged)

I'm more comfortable with the above interpretation of how to handle reflection in code.

Secondly, shouldn't the AvoidAccessibilityAlteration check be only for setAccessible(true) since false forces the JVM to enforce the language semantics?

Adding StackOverflow Q&A for context:

[How to limit setAccessible to only "legitimate" uses](https://stackoverflow.com/questions/2481862/how-to-limit-setaccessible-to-only-legitimate-uses)

**Comment by pzygielo** [comment]
I've a question, which I feel is kind of related, and I'm not sure it deserves separate discussion.
But GH has only field for _Answer_ in this discussion, hence I don't know if this is proper place too, as this is not an answer.

------

For the code
```java
import java.security.PrivilegedAction;

public class AvoidAccessibilityAlteration {
  public void f(String cn) throws Exception {
    Class.forName(cn).getDeclaredConstructor().newInstance();
  }
}
```
the rule under discussion is found as violated by current PMD (6.37). `PrivilegedAction` is not used, but `getDeclaredConstructor` is.
Once the import is removed - the violation is gone as well. I don't understand the description or the XPath expression (or both).

Shouldn't used method `getDeclaredConstructor` be reported anyway, regardless of import present or not?

Seems the rule was initially *controversial*, but is no more.

**Comment by adangel** [comment]
> Can you explain what's the purpose of the above rule?

The rule wants to flag cases, where you e.g. make a private field accessible via reflection. As described within the rule description: "This violates the principle of encapsulation."

It doesn't not do a good job though, e.g. just calling `getDeclaredConstructors` does do any harm yet. But calling a private constructor by making it accessible via `setAccessible` is problematic.

> Why is the rule restricted only to constructors and not methods and fields?

Because the original author didn't think further. He/she probably just had a specific problem with someone calling a private constructor...
but checking for `setAccessible` should be enough anyway (although the rule will find any methods, that are called `setAccessible` instead of just calls to `java.lang.reflect.AccessibleObject.setAccessible(boolean)`. Note, that there is also a static method `java.lang.reflect.AccessibleObject.setAccessible(AccessibleObject[], boolean)`.

> Why is the rule enforced despite the presence of a PrivilegedAction block?

That's probably a simple bug and the rule is just untestet: It has not a single test case.
I think, checking for PrivilegedAction doesn't make sense - it would just wrap e.g. a call to `setAccessible`, which would then be found anyways.

I tend to agree with your view of PrivilegedAction, that it is more the opposite: If the programmer used PrivilegedAction, we could assume, that he thought through it and deliberately wants to access a private field. So, using PrivilegedAction could be used as a marker for exceptions to the rule.

> Secondly, shouldn't the AvoidAccessibilityAlteration check be only for setAccessible(true) since false forces the JVM to enforce the language semantics?

Maybe, what why would you write `setAccessible(false)` in the first place? It's maybe better to flag all places, where you mess with accessibility...


> the rule under discussion is found as violated by current PMD (6.37). PrivilegedAction is not used, but getDeclaredConstructor is.
Once the import is removed - the violation is gone as well. I don't understand the description or the XPath expression (or both).

That's the problem, that the rule is not tested at all - the original author wanted to flag all uses of `setAccessible`, `getDeclaredConstructors` *and* `PrivilegedActions` - but in order to find all, you need to OR them, or - if you use 2 xpath queries as the author did - you need to create a union.

The corrected XPath would look like this:

```
(: first XPath query: method calls to getDeclaredConstructor(s), setAccessible :)
//PrimaryExpression
[
  PrimarySuffix[
        ends-with(@Image,'getDeclaredConstructors')
                or
        ends-with(@Image,'getDeclaredConstructor')
                or
        ends-with(@Image,'setAccessible')
        ]
or
 PrimaryPrefix/Name[
        ends-with(@Image,'getDeclaredConstructor')
        or
        ends-with(@Image,'getDeclaredConstructors')
        or
        starts-with(@Image,'AccessibleObject.setAccessible')
        ]
 ]

| (: that's the union operator :)

(: second XPath query: any import of PrivilegedAction :)
//ImportDeclaration/Name[
        contains(@Image,'java.security.PrivilegedAction')]
```

But as described earlier, the PrivilegedAction part should probably be inverted, so that only usages without PrivilegedActions are flagged....
